_source/_posts/2021-12-18-native-java-helidon.adoc

```diff
@@ -0,0 +1,403 @@
+---
+layout: blog_post
+title: "Build REST APIs and Native Java Apps with Helidon"
+author: matt-raible
+by: advocate
+communities: [java]
+description: "Learn how to create REST APIs and native images with Helidon, a Java microservices framework from Oracle."
+tags: [native, java, graalvm, helidon]
+tweets:
+- ""
+- ""
+- ""
+image:
+type: conversion
+github: https://github.com/oktadev/native-java-examples
+---
+:page-liquid:
+:toc: macro
+:experimental:
+
+Project Helidon is a set of Java libraries for writing microservices. It's an open source, Apache 2.0-licensed, Oracle-sponsored project and contains support for MicroProfile as well as a reactive, functional API. Helidon's two different style are known as Helidon MP and SE, respectively. The https://helidon.io[helidon.io] website says:
+
+====
+Since Helidon is simply a collection of Java libraries running on a fast Netty core, there is no extra overhead or bloat.
+====
+
+image::{% asset_path 'blog/native-java-helidon/helidon-mp.png' %}[alt=Helidon supports MicroProfile,width=400,align=center]
+
+Today, I'm going to put this statement to the test! First, I'll show you how to run a secure, OAuth 2.0-protected, Java REST API that allows JWT authentication. Then, I'll show you how to build a native image with GraalVM. Finally, I'll compare Helidon to the most popular Java REST API frameworks: Micronaut, Quarkus, and Spring Book.
+
+My link:/blog/2021/06/18/native-java-framework-comparison[Build Native Java Apps with Micronaut, Quarkus, and Spring Boot] post compares startup times and memory usage of these three frameworks. How does Helidon perform? Read on!
+
+**Prerequisites**
+
+- https://sdkman.io/[SDKMAN]
+- https://maven.apache.org[Maven]
+- https://httpie.io/[HTTPie] (a better version of cURL)
+- An https://developer.okta.com[Okta Developer] Account
+
+toc::[]
+
+== Get Started with Helidon
+
+I created a GitHub repository you can clone and run to get started with Helidon quickly.
+
+[source,shell]
+----
+git clone https://github.com/oktadev/native-java-examples.git
+cd native-java-examples/helidon
+----
+
+I'll show you how I created this example in a section below. First, let's see how to run it.
+
+Open the `helidon` directory in your favorite IDE, so you have easy access to the example's project files.
+
+TIP: If you just want to see how to build a native image, skip to the link:#build-native-images-for-helidon[build native images for Helidon] section.
+
+== Install a JDK with GraalVM
+
+You will need a JDK with GraalVM and its `native-image` compiler. Using SDKMAN, run the following command and set it as the default:
+
+[source,shell]
+----
+sdk install java 21.3.0.r17-grl
+----
+
+Add the native extension to the JDK:
+
+[source,shell]
+----
+gu install native-image
+----
+
+== Run a Helidon Java API
+
+In a terminal window, run `mvn package` to package the app into a JAR. Then use `java -jar` to run it.
+
+[source,shell]
+----
+mvn package
+java -jar target/helidon.jar
```

Comment:
You can also run `mvn exec:exec`
or set the default goal in your project pom: `<defaultGoal>exec:exec</defaultGoal>`

Pet peeve, one of the Helidon parent poms uses an older version of the `exec-maven-plugin`, ideally, this _would_ use `exec:java` instead, but I'm splitting hairs, and they probably had a reason for it 😄 

I was also able to run and debug the app through IntelliJ by creating a run configuration with `io.helidon.microprofile.cdi.Main`

Reply:
I don't have an opportunity to try this right now. Do you think I should add it or just leave it out?